### PR TITLE
Schedule only one CleanOutServer job. Introduced new Drain phase.

### PR DESCRIPTION
**Pull request by maierlars**

### CODE REVIEW

File: pkg/apis/deployment/v1alpha/member_phase.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ const (
 	MemberPhaseFailed MemberPhase = "Failed"
 	// MemberPhaseCleanOut indicates that a dbserver is in the process of being cleaned out
 	MemberPhaseCleanOut MemberPhase = "CleanOut"
+	// MemberPhaseDrain indicates that a dbserver is n the process of being cleaned out as result of draining a node
```
Comment (ewoutp) [inline pkg/apis/deployment/v1alpha/member_phase.go]: `n` -> `in`

Comment (ewoutp) [inline pkg/apis/deployment/v1alpha/member_phase.go]: Suggest to add a helper `func (p MemberPhase) IsCleanOutOrDrain() bool`
will be cleaner in the rest of the code.

File: pkg/deployment/reconcile/plan_builder.go [hunk 2]
```diff
@@ -120,7 +120,7 @@ func createPlan(log zerolog.Logger, apiObject k8sutil.APIObject,
 
 	// Check for cleaned out dbserver in created state
 	for _, m := range status.Members.DBServers {
-		if len(plan) == 0 && m.Phase == api.MemberPhaseCreated && m.Conditions.IsTrue(api.ConditionTypeCleanedOut) {
+		if len(plan) == 0 && (m.Phase == api.MemberPhaseCreated || m.Phase == api.MemberPhaseDrain) && m.Conditions.IsTrue(api.ConditionTypeCleanedOut) {
```
Comment (ewoutp) [inline pkg/deployment/reconcile/plan_builder.go]: Use `IsCleanOutOrDrain` (or use brackets to clarify OR or AND preference)

File: pkg/deployment/resources/pod_finalizers.go [hunk 3]
```diff
@@ -123,7 +123,7 @@ func (r *Resources) inspectFinalizerPodDrainDBServer(ctx context.Context, log ze
 	}
 
 	// If this DBServer is cleaned out, we need to remove the PVC.
-	if memberStatus.Conditions.IsTrue(api.ConditionTypeCleanedOut) {
+	if memberStatus.Conditions.IsTrue(api.ConditionTypeCleanedOut) || memberStatus.Phase == api.MemberPhaseDrain {
```
Comment (ewoutp) [inline pkg/deployment/resources/pod_finalizers.go]: Use `IsCleanOutOrDrain`

File: pkg/deployment/resources/pod_termination.go [hunk 4]
```diff
@@ -185,15 +187,58 @@ func (r *Resources) prepareDBServerPodTermination(ctx context.Context, log zerol
 		return nil
 	}
 	// Not cleaned out yet, check member status
-	if memberStatus.Conditions.IsTrue(api.ConditionTypeTerminated) {
+	if memberStatus.Conditions.IsTrue(api.ConditionTypeTerminated) || (memberStatus.Phase == api.MemberPhaseCreated && !k8sutil.IsPodReady(p)) {
```
Comment (ewoutp) [inline pkg/deployment/resources/pod_termination.go]: Why this additional expression?

File: pkg/deployment/resources/pod_termination.go [hunk 5]
```diff
@@ -185,15 +187,58 @@ func (r *Resources) prepareDBServerPodTermination(ctx context.Context, log zerol
...
 		log.Warn().Msg("Member is already terminated before it could be cleaned out. Not good, but removing dbserver pod because we cannot do anything further")
+		// At this point we have to set CleanedOut to true,
+		// because we can no longer reason about the state in the agency and
+		// bringing back the dbserver again may result in an cleaned out server without us knowing
+		memberStatus.Conditions.Update(api.ConditionTypeCleanedOut, true, "Draining server failed", "")
+		memberStatus.Phase = api.MemberPhaseCreated
```
Comment (ewoutp) [inline pkg/deployment/resources/pod_termination.go]: Shouldn't this be condition?
You have the risk of overriding unwanted phases.

File: pkg/deployment/resources/pod_termination.go [hunk 6]
```diff
@@ -185,15 +187,58 @@ func (r *Resources) prepareDBServerPodTermination(ctx context.Context, log zerol
...
+		}
+		if jobStatus.IsFailed() {
+			log.Warn().Str("reason", jobStatus.Reason()).Msg("Cleanout Job failed. Aborting plan")
+			// Revert cleanout state
+			memberStatus.Phase = api.MemberPhaseCreated
+			memberStatus.CleanoutJobID = ""
+			return maskAny(fmt.Errorf("Clean out server job failed"))
+		}
+	} else {
+		log.Warn().Msgf("Unexpected MemberPhase %s, allow removal", memberStatus.Phase)
```
Comment (ewoutp) [inline pkg/deployment/resources/pod_termination.go]: Is this really what should happen  if `phase == MemberPhaseCleanOut`?

Comment (lamai93) [reply pkg/deployment/resources/pod_termination.go]: Probably not

**Pull request merged**